providers/mlx5/dr_send.c

```diff
@@ -0,0 +1,1039 @@
+	wq_size = max(wq_size, MLX5_SEND_WQE_BB);
+	dr_qp->rq.wqe_cnt = wq_size / wqe_size;
+	dr_qp->rq.wqe_shift = mlx5_ilog2(wqe_size);
+	dr_qp->rq.max_post = 1 << mlx5_ilog2(wq_size / wqe_size);
+	dr_qp->rq.max_gs = wqe_size / sizeof(struct mlx5_wqe_data_seg);
+
+	return wq_size;
+}
+
+static int dr_calc_wq_size(struct dr_qp *dr_qp, struct dr_qp_init_attr *attr)
+{
+	int result;
+	int ret;
+
+	result = dr_calc_sq_size(dr_qp, attr);
+
+	ret = dr_calc_rq_size(dr_qp, attr);
+
+	result += ret;
+	dr_qp->sq.offset = ret;
+	dr_qp->rq.offset = 0;
+
+	return result;
+}
+
+static int dr_qp_alloc_buf(struct dr_qp *dr_qp, int size)
+{
+	int al_size;
+	int ret;
+
+	dr_qp->sq.wqe_head = malloc(dr_qp->sq.wqe_cnt *
+				    sizeof(*dr_qp->sq.wqe_head));
+	if (!dr_qp->sq.wqe_head) {
+		errno = ENOMEM;
+		return errno;
+	}
+
+	al_size = align(size, sysconf(_SC_PAGESIZE));
+	ret = posix_memalign(&dr_qp->buf.buf, sysconf(_SC_PAGESIZE), al_size);
+	if (ret) {
+		errno = ret;
+		goto free_wqe_head;
+	}
+
+	dr_qp->buf.length = al_size;
+	dr_qp->buf.type = MLX5_ALLOC_TYPE_ANON;
+	memset(dr_qp->buf.buf, 0, dr_qp->buf.length);
+
+	return 0;
+
+free_wqe_head:
+	free(dr_qp->sq.wqe_head);
+	return ret;
+}
+
+static struct dr_qp *dr_create_rc_qp(struct ibv_context *ctx,
+				     struct dr_qp_init_attr *attr)
+{
+	struct dr_devx_qp_create_attr qp_create_attr;
+	struct mlx5dv_devx_obj *obj;
+	struct dr_qp *dr_qp;
+	int size;
+	int ret;
+
+	dr_qp = calloc(1, sizeof(*dr_qp));
+	if (!dr_qp) {
+		errno = ENOMEM;
+		return NULL;
+	}
+
+	size = dr_calc_wq_size(dr_qp, attr);
+
+	if (dr_qp_alloc_buf(dr_qp, size))
+		goto err_alloc_bufs;
+
+	dr_qp->sq_start = dr_qp->buf.buf + dr_qp->sq.offset;
+	dr_qp->sq.qend = dr_qp->buf.buf + dr_qp->sq.offset +
+		(dr_qp->sq.wqe_cnt << dr_qp->sq.wqe_shift);
+	dr_qp->rq.head = 0;
+	dr_qp->rq.tail = 0;
+	dr_qp->sq.cur_post = 0;
+
+	ret = posix_memalign((void **)&dr_qp->db, 8, 8);
+	if (ret) {
+		errno = ret;
+		goto err_db_alloc;
+	}
+
+	dr_qp->db[MLX5_RCV_DBR] = 0;
+	dr_qp->db[MLX5_SND_DBR] = 0;
+	dr_qp->db_umem = mlx5dv_devx_umem_reg(ctx, dr_qp->db, 8,
+					      IBV_ACCESS_LOCAL_WRITE |
+					      IBV_ACCESS_REMOTE_WRITE |
+					      IBV_ACCESS_REMOTE_READ);
+	if (!dr_qp->db_umem)
+		goto err_db_umem;
+
+	dr_qp->buf_umem = mlx5dv_devx_umem_reg(ctx, dr_qp->buf.buf,
+					       dr_qp->buf.length,
+					       IBV_ACCESS_LOCAL_WRITE |
+					       IBV_ACCESS_REMOTE_WRITE |
+					       IBV_ACCESS_REMOTE_READ);
+	if (!dr_qp->buf_umem)
+		goto err_buf_umem;
+
+	qp_create_attr.page_id		= attr->uar->page_id;
+	qp_create_attr.pdn		= attr->pdn;
+	qp_create_attr.cqn		= attr->cqn;
+	qp_create_attr.pm_state		= MLX5_QPC_PM_STATE_MIGRATED;
+	qp_create_attr.service_type	= MLX5_QPC_ST_RC;
+	qp_create_attr.buff_umem_id	= dr_qp->buf_umem->umem_id;
+	qp_create_attr.db_umem_id	= dr_qp->db_umem->umem_id;
+	qp_create_attr.sq_wqe_cnt	= dr_qp->sq.wqe_cnt;
+	qp_create_attr.rq_wqe_cnt	= dr_qp->rq.wqe_cnt;
+	qp_create_attr.rq_wqe_shift	= dr_qp->rq.wqe_shift;
```

Comment:
I would be happy if much this code were run through clang-format to get rid of stuff like this..